torch/distributed/_shard/api.py

```diff
@@ -118,3 +123,80 @@ def shard_parameter(
+
+
+class ShardedModuleSwapper(abc.ABC):
+    @abc.abstractmethod
+    def process(self, module: nn.Module) -> nn.Module:
+        """
+        Processes a module and if needed swaps it with a custom sharded
+        Implementation. Should return ``None`` if no swapping should be
+        performed.
+
+        The passed in module would consist of ShardedTensors and a common way
+        to perform module swapping would be to use the state_dict of the passed
+        in module and apply it to the new sharded module via its
+        load_state_dict method.
+        """
+        pass
+
+def shard_module(
+    module: nn.Module,
+    plan: ShardingPlan,
+    sharded_module_swapper: ShardedModuleSwapper = None,
+    reshard_output: ShardingPlan = None,
+    src_rank=0,
+    process_group=None
+):
+    """
+    Shards a given module according to the provided sharding_plan. This method
+    first shards all the parameters according to the given sharding_plan. Next,
+    If sharded_module_swapper is specified, it recursively traverses the module
+    tree and calls ``sharded_module_swapper.process()`` for each submodule.
+    If ``sharded_module_swapper.process()`` returns ``None``, the method continues
+    to recurse further down the tree. If ``sharded_module_swapper.process()``
+    returns an ``nn.Module``, then the current module is replaced with the return
+    value and we don't recurse further.
+
+    If ``reshard_output`` is specified, the module's output is resharded
+    according to the provided sharding plan.
+
+    Args:
+        module: The module to apply sharding to
+        sharding_plan: Dict from param name to PlacementSpec/ShardingSpec
+        to apply to each parameter.
+        sharded_module_swapper: Implementation of ShardedModuleSwapper for module
+        swapping.
+        reshard_output: ShardingPlan in terms of how
+        we would like to reshard the output of a module which produces a
+        ShardedTensor. It can also allow gathering the local output for each
+        rank as a Tensor to ensure further processing in a data parallel
+        fashion.
+    """
+    if sharded_module_swapper is not None:
+        raise NotImplementedError("custom module swapping not implemented yet!")
+    if reshard_output is not None:
+        raise NotImplementedError("output resharding not implemented yet!")
+
+    for mod_prefix, mod in module.named_modules():
+        # memo is used to avoid duplicate params between modules, this logic
+        # is mostly copied from the module._named_members(), with adaptions
+        # to handle parameter sharding.
+        memo = set()
+        # create a list from the dict keys, because we are mutating the
+        # parameters on the fly, we need to use a separate list instead.
+        param_keys = list(mod._parameters.keys())
+        for k in param_keys:
+            v = mod._parameters[k]
+            name = mod_prefix + ('.' if mod_prefix else '') + k
```

Comment:
N00b question, is this expected? `mod_prefix.mod_prefix`? But why we don't check `mod_prefix` for the first one?

Reply:
can you elaborate more about what's `mod_prefix.mod_prefix`? we are just concating a module prefix together with its parameter name.